### ISSUE TITLE
FP no-invalid-await on JS file in SonarLint

### ISSUE THREAD

**Issue by peterpoliwoda** [SonarSource/SonarJS#2636]
Reflecting on a forum issue spoken about here:
https://community.sonarsource.com/t/refactor-this-redundant-await-on-a-non-promise-is-not-correct/28800

My VSCode extension wrongly tells me that I don't need to await for a non-promise. Problem is, my function is an async function and by all means does return a promise as per ES6 standards.

```JavaScript
async function getAuthenticationToken() {
    try {
        const tokenUri = `${keycloakAuthServerUrl}realms/${keycloakRealm}/protocol/openid-connect/token`;

        const params = new URLSearchParams();
        params.append('client_id', DB_KEYCLOAK_CLIENT_ID);
        params.append('client_secret', DB_KEYCLOAK_CLIENT_SECRET);
        params.append('grant_type', MY_GRANT_TYPE);

        const { data } = await axios.post(
            tokenUri,
            params,
            {
                headers: {
                    'Content-Type': 'application/x-www-form-urlencoded'
                }
            }
        );
        return data.access_token;
    } catch (err) {
        logger.error('[getAuthenticationToken] Error getting authentication token', err);
        throw err;
    }
}

async function insertToMd4DatabaseApi(dbTable, data) {
    logger.info('[insertToMd4DatabaseApi] Insert data to Database API', { dbTable });
    logger.debug('[insertToMd4DatabaseApi] Data being saved', { dbTable, data });
    try {
        const authToken = await getAuthenticationToken(); // This part gets underlined by Sonar lint
        return axios.post(`/${dbTable}`, data, {
            baseURL: apiBaseUri,
            headers: {
                'some-auth-stuff': `Bearer ${authToken}`
            }
        });
    } catch (err) {
        logger.error('[insertToMd4DatabaseApi] Error saving data to Database API', err);
        throw err;
    }
}
```

This rule is causing some stir in my code and has caused some bugs. Please add support for async functions in JavaScript as this is a bug in linting at the moment.

**Comment by vilchik-elena** [comment]
Hi @peterpoliwoda, 

It seems I rushed to add the ticket to the milestone as I can't reproduce it. Could you share self contained code snippet reproducing FP?

**Comment by peterpoliwoda** [comment]
Hi @vilchik-elena I've updated the code above to reflect the two actual functions in case the code wasn't showing up for you. I updated to show that my underlined code is also wrapped within a try/catch block which might be the reason it's not handled. Please try now.

**Comment by vilchik-elena** [comment]
I still fail to reproduce, even with `axios` installed.  Are you using latest plugin? 
Could you may be try to set up a separate project (so that dependencies are set)?

**Comment by peterpoliwoda** [comment]
I found there was a single quote after `MY_GRANT_TYPE'` which shouldn't be there. Code is fixed above. 

![image](https://user-images.githubusercontent.com/14016749/120494965-2c517e80-c3b4-11eb-9f3c-54c53314ad15.png)

We are both speaking about the SonarLint VSCode Plugin BTW, right?
![image](https://user-images.githubusercontent.com/14016749/120495763-e812ae00-c3b4-11eb-8687-36067b7805d3.png)

I've set up a project with a single file where it also gives out here:
https://github.com/peterpoliwoda/test-sonarlint

**Comment by search-acumen** [comment]
![image](https://user-images.githubusercontent.com/73835758/120616207-54df8400-c459-11eb-9118-cda7055d0651.png)

Same issue here using SonarLint VSCode Plugin 2.0.0

**Comment by vilchik-elena** [comment]
thanks for reproducer! having it on my machine as well, investigating why I can't reproduce the same in unit test

**Comment by search-acumen** [comment]
If it helps, it seems to only happen if the async function is in the same js file as the await. Awaits to imported async functions don't seem to throw the warning.

**Comment by codyebberson** [comment]
Here is another example, no dependencies, only 8 lines:

```typescript
interface MyQuery<T> extends Pick<Promise<T>, keyof Promise<T>> {
  toQuery(): string;
}

export async function demo(query: MyQuery<string>) {
  const result = await query;
  console.log(result);
}
```

![image](https://user-images.githubusercontent.com/749094/120696084-fae2bc80-c460-11eb-8636-b85eadeb30fb.png)

This is effectively what [Knex](https://github.com/knex/knex) is doing, which is a pretty popular SQL builder:

https://github.com/knex/knex/blob/dfbc76b915fff9f40e806ac2e0aafe8e4e5b6965/types/index.d.ts#L1742

```typescript
interface ChainableInterface<T = any> extends Pick<Promise<T>, keyof Promise<T> & ExposedPromiseKeys>, StringTagSupport {
  generateDdlCommands(): Promise<{ pre: string[], sql: string[], post: string[] }>;
  toQuery(): string;
```

**Comment by vilchik-elena** [comment]
It seems related to the way we do analysis of JavaScript files using TypeScript compiler. Hope to fix it soon

**Comment by vilchik-elena** [comment]
Finally found the root problem, we are missing type information because in SonarLint context we are not able to generate tsconfig file (https://github.com/SonarSource/SonarJS/blob/1b593bddb6270e79076fe15f9151003a586f33ff/sonar-javascript-plugin/src/main/java/org/sonar/plugins/javascript/eslint/TsConfigProvider.java#L141-L145). 

We want to do a major rework so that this generated tsconfig file would not be required but until that we can provide a simple fix to not report if type of awaited argument is unknown.

**Comment by peterpoliwoda** [comment]
Wow, that's completely unrelated then hah. Of course that was the cause, what else haha. Thank you for looking into that Elena 🐺 !

**Comment by Tranthanh98** [comment]
I have a same issue, my function obviously return a Promise but sonarqube show this error
<img width="898" alt="Screen Shot 2021-11-17 at 13 20 24" src="https://user-images.githubusercontent.com/51749033/142145483-280ba717-d4fd-470f-bc4c-70de2a572977.png">

**Comment by tr3ysmith** [comment]
Same issue here, is there something we have to update in VSCode to make this fix active?

**Comment by vilchik-elena** [comment]
@tr3ysmith yeah. If you are not using connected mode, then update SonarLint plugin

**Comment by kamesh95** [comment]
@vilchik-elena  We are using SonarQube Enterprise v8.9.0 and still this issue shows up on our dashboard as a Code Smell. Can you let me know with which version of SonarQube this false positive is not detected? Thanks.

**Comment by vilchik-elena** [comment]
@kamesh95 you should update to latest SQ (9.2 as of now)

**Comment by robertpatrick** [comment]
This issue is a pain for our JavaScript project.  Basically, any async function call always returns a Promise--even if the code does not indicate it.  I have to turn this rule off because its implementation is naive and results in thousands of false positives...

**Comment by VinWare** [comment]
The rule still gives an issue for TS files (typescript:S4123). Is the mentioned fix supposed to work for TS or must it be done separately. Happy to raise a new issue if that is appropriate here.

**Comment by vishalamrutiyacodemonk** [comment]
Which version of SQ resolves this issue? Currently, I installed 9.6.1 version

**Comment by hybridx** [comment]
Why is this issue close? This is still a bug in the latest (SonarLint v3.17.0)

**Comment by kamal-brill** [comment]
We still see it even in "Version 9.9.4 (build 87374)". Is there any update on this ?

**Comment by Gean-Vasconez** [comment]
This issue persists on SonarQube for IDE 10.20.0.80985 in JetBrains Rider. Is the fix only for VSCode?
